### PR TITLE
[1542][ADD] base_update_restrict_model

### PR DESCRIPTION
**Pull request by TimLai125**
[1542](https://www.quartile.co/web#id=1542&model=project.task&view_type=form&menu_id=)

### CODE REVIEW

File: base_update_restrict_model/tests/test_base_update_restrict_model.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        ]
+        cls.partner_model.update_restrict = True
+        cls.test_partner = cls.env["res.partner"].create({"name": "Test Partner"})
+        cls.restrict_test_user = cls.env["res.users"].create(
+            {
+                "name": "Resticted user",
+                "login": "resticted@example.com",
+                "restrict_update_permit": False,
+            }
+        )
```
Comment (yostashiro) [inline base_update_restrict_model/tests/test_base_update_restrict_model.py]: Looks like Travis is complaining that there is no email for the partner. Would it make it simpler to just directly create the user without creating a partner beforehand. I suppose that email should be automatically updated with the login that way.

File: base_update_restrict_model/views/res_users_views.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        <field name="arch" type="xml">
+            <field name="company_id" position="after">
+                <field name="restrict_update_permit" invisible="1" />
+            </field>
+            <div class="oe_button_box" name="button_box" position="inside">
+                <button
+                    name="set_restrict_update_permit"
+                    type="object"
+                    class="oe_stat_button"
+                    icon="fa-eye"
```
Comment (yostashiro) [inline base_update_restrict_model/views/res_users_views.xml]: These icons are a bit misleading (making an object read-only and restricting updates are different things).

I also prefer this style: https://github.com/odoo/odoo/blob/2fb44a86819fc6a7a6507d120c79639e2e9f8a6b/odoo/addons/base/views/res_users_views.xml#L166-L168

**Review by yostashiro — CHANGES_REQUESTED** [review]
I suggest that we add add USAGE.rst (and maybe CONFIGURE.rst) to explain the settings/how to use it.

I understand that this module currently supports the global restriction (i.e. restrictions are either applied to all the selected models or otherwise per user) only, which itself is fine when we do not consider further granularity. When we start thinking about control per user per model, for example, the current design may make it a bit difficult to do so without compromising the UI/UX.

Ideally, the optimal design I think is to have another model ('model.access.restriction' or something) with a dedicated menu item, and contain the global and per-user restrictions in there, kind of similar to how record rules are structured.

It depends on how we envision the final state of this module, so please give me your idea.

File: base_update_restrict_model/models/res_users.py [hunk 3]
```diff
@@ -8,13 +8,10 @@ class ResUsers(models.Model):
     _inherit = "res.users"
 
     restrict_update_permit = fields.Boolean(
```
Comment (yostashiro) [inline base_update_restrict_model/models/res_users.py]:
```suggestion
    unrestrict_model_update = fields.Boolean(
```

File: base_update_restrict_model/models/res_users.py [hunk 4]
```diff
@@ -8,13 +8,10 @@ class ResUsers(models.Model):
...
-        "Update Permit", help="Set True Means This User can update restricted model."
+        "Update Permit", help="Set to true and the user can update restricted model."
```
Comment (yostashiro) [inline base_update_restrict_model/models/res_users.py]:
```suggestion
        "Unrestrict Model Update", help="Set to true and the user can update restricted model."
```

File: base_update_restrict_model/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2021 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Update Restrict Model",
+    "version": "12.0.1.0.0",
+    "depends": ["base"],
+    "website": "https://github.com/OCA/server-tools",
+    "author": "Odoo Community Association (OCA), Quartile Limited",
```
Comment (yostashiro) [inline base_update_restrict_model/__manifest__.py]: Doesn't a company usually come before OCA? Not sure if that matters.

File: base_update_restrict_model/models/ir_model.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2021 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import fields, models
+
+
+class IrModel(models.Model):
+    _inherit = "ir.model"
+
+    update_restrict = fields.Boolean("Update Restrict Model")
```
Comment (yostashiro) [inline base_update_restrict_model/models/ir_model.py]:
```suggestion
    restrict_update = fields.Boolean(help="When selected, the model is restricted to read-only unless the user has the special permission.")
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
Please follow up on the point of crediting Moises as we talked.

base_model_restrict_update sounds better to me as a module name, but it's not a strong suggestion.

Made miscellaneous comments on non-essential points.  I leave it up to you how you treat them.

I'd suggest that we always create a PR in the OCA first. We miss the opportunity to receive the community feedback at the initial stage, otherwise.

File: base_model_restrict_update/views/res_users_views.xml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="view_users_form" model="ir.ui.view">
+        <field name="name">res.users.form</field>
+        <field name="model">res.users</field>
+        <field name="inherit_id" ref="base.view_users_form" />
+        <field name="arch" type="xml">
+            <div class="oe_button_box" name="button_box" position="inside">
+                <button name="toggle_unrestrict_model_update" type="object" class="oe_stat_button" icon="fa-eye">
```
Comment (yostashiro) [inline base_model_restrict_update/views/res_users_views.xml]: I think `fa-pencil` may better represent what it is.

**Review by yostashiro — COMMENTED** [review]
@TimLai125 Please accredit Moises.

**Review by yostashiro — APPROVED** [review]
LGTM 👍

**Comment by TimLai125** [comment]
taken over by https://github.com/qrtl/hls-oca/pull/1

**Pull request closed without merging**